ocaml/xapi/xapi_psr.ml

```diff
@@ -0,0 +1,157 @@
+      |> (* remove leading and trailing '"' *)
+      fun s -> String.sub s 1 (String.length s - 2)
+
+    let checkpoint_of_string s =
+      let rpc = Rpc.rpc_of_string s in
+      match Rpcmarshal.unmarshal checkpoint.Rpc.Types.ty rpc with
+      | Ok x ->
+          x
+      | Error _ ->
+          raise (Cannot_parse_checkpoint s)
+
+    let rec iter_break f = function
+      | [] ->
+          Ok ()
+      | x :: xs ->
+          f x >>= fun _ -> (iter_break [@tailcall]) f xs
+
+    let rec go pool_secrets master members = function
+      | No_checkpoint ->
+          (* if we fail to backup the pool secrets, it doesn't really matter,
+             you can simply restart the rotation *)
+          Impl.backup pool_secrets ;
+          (go [@tailcall]) pool_secrets master members Accept_new_pool_secret
+      | Accept_new_pool_secret ->
+          Impl.save_checkpoint (string_of_checkpoint Accept_new_pool_secret) ;
+          master :: members
+          |> iter_break (fun host ->
+                 try
+                   Impl.tell_accept_new_pool_secret pool_secrets host ;
+                   Ok ()
+                 with e ->
+                   D.error
+                     "failed while telling host to accept new pool secret. \
+                      error= %s"
+                     (Printexc.to_string e) ;
+                   Error (Failed_during_accept_new_pool_secret, host))
+          >>= fun () ->
+          (go [@tailcall]) pool_secrets master members Send_new_pool_secret
+      | Send_new_pool_secret ->
+          Impl.save_checkpoint (string_of_checkpoint Send_new_pool_secret) ;
+          master :: members
+          |> iter_break (fun host ->
+                 try
+                   Impl.tell_send_new_pool_secret pool_secrets host ;
+                   Ok ()
+                 with e ->
+                   D.error
+                     "failed while telling hosts to send new pool secret. \
+                      error= %s"
+                     (Printexc.to_string e) ;
+                   Error (Failed_during_send_new_pool_secret, host))
+          >>= fun () ->
+          (go [@tailcall]) pool_secrets master members Cleanup_members
+      | Cleanup_members ->
+          Impl.save_checkpoint (string_of_checkpoint Cleanup_members) ;
+          members
+          |> iter_break (fun member ->
+                 try
+                   Impl.tell_cleanup_old_pool_secret member ;
+                   Ok ()
+                 with e ->
+                   D.error "failed while telling hosts to cleanup. error= %s"
+                     (Printexc.to_string e) ;
+                   Error (Failed_during_cleanup_member, member))
+          >>= fun () ->
+          (go [@tailcall]) pool_secrets master members Cleanup_master
+      | Cleanup_master ->
+          Impl.save_checkpoint (string_of_checkpoint Cleanup_master) ;
+          Helpers.retry_forever ~interval:1. ~doc:"cleanup master after PSR"
```

Comment:
This means that the API call will potentially hang forever, and the caller will not be able to cancel it. What is the consequence of this failure state? Everything works with the new token, but the master still accepts the old one? It may be better to treat it like the other failure cases: fail with and error describing the situation and let the user retry the call.